### PR TITLE
Pass all echo-* tests in basic feature tests

### PR DESCRIPTION
**Pull request by yzhang90**
* support map type argument

### CODE REVIEW

**Comment by hjorthjort** [comment]
What does `echo-*` test mean?

File: Jenkinsfile [hunk 1]
```diff
@@ -17,7 +17,7 @@ pipeline {
       }
     }
     stage('Test') {
-      options { timeout(time: 10, unit: 'MINUTES') }
+      options { timeout(time: 15, unit: 'MINUTES') }
```
Comment (hjorthjort) [inline Jenkinsfile:20]: If the average run time is 10+ mins, you can bump this to 20 mins to give us some margin

**Comment by yzhang90** [comment]
> What does `echo-*` test mean?

all the tests start with echo
```
deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_array_u8.scen.json
deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_big_int.scen.json
deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_big_uint.scen.json
deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_boxed_bytes.scen.json
deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_i32.scen.json
deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_i64.scen.json
deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_nothing.scen.json
deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_ser_ex_1.scen.json
deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_slice_u8.scen.json
deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_str_box.scen.json
deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_string.scen.json
deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_str.scen.json
deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_u64.scen.json
deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_usize.scen.json
deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_varags_tuples.scen.json
deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_varargs_u32.scen.json
deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_vec_u8.scen.json
```

File: elrond.md [hunk 2]
```diff
@@ -961,7 +977,13 @@ TODO: Implement [reserved keys and read-only runtimes](https://github.com/Elrond
 ```k
     // extern long long smallIntGetUnsignedArgument(void *context, int32_t id);
     rule <instrs> hostCall("env", "smallIntGetUnsignedArgument", [ i32 .ValTypes ] -> [ i64 .ValTypes ])
-               => #returnIfUInt64(Bytes2Int({ARGS[ARG_IDX]}:>Bytes, BE, Unsigned)) ... </instrs>
+               => #returnIfUInt64(Bytes2Int({ARGS[ARG_IDX]}:>Bytes, BE, Unsigned), "argument out of range") ... </instrs>
```
Comment (hjorthjort) [inline elrond.md:980]: "argument out of range" sounds like you tried to access an argument with index too high, to me.
How about "argument " +String Int2String(ARG_IDX) +String " can't be converted to an unsigned i64"

File: elrond.md [hunk 3]
```diff
@@ -961,7 +977,13 @@ TODO: Implement [reserved keys and read-only runtimes](https://github.com/Elrond
...
+         <locals> 0 |-> <i32> ARG_IDX </locals>
+         <callArgs> ARGS </callArgs>
+
+    // extern long long smallIntGetSignedArgument(void *context, int32_t id);
+    rule <instrs> hostCall("env", "smallIntGetSignedArgument", [ i32 .ValTypes ] -> [ i64 .ValTypes ])
+               => #returnIfSInt64(Bytes2Int({ARGS[ARG_IDX]}:>Bytes, BE, Signed), "argument out of range") ... </instrs>
```
Comment (hjorthjort) [inline elrond.md:986]: Same here, but "signed"

File: elrond.md [hunk 4]
```diff
@@ -989,7 +1018,7 @@ TODO: Implement [reserved keys and read-only runtimes](https://github.com/Elrond
     rule <instrs> hostCall("env", "smallIntStorageLoadUnsigned", [ i32 i32 .ValTypes ] -> [ i64 .ValTypes ])
                => #memLoad(KEYOFFSET, KEYLENGTH)
                ~> #storageLoad
-               ~> #loadBytesAsUInt64
+               ~> #loadBytesAsUInt64("argument out of range")
```
Comment (hjorthjort) [inline elrond.md:1021]: Same here

**Review by hjorthjort — CHANGES_REQUESTED** [review]

**Comment by hjorthjort** [comment]
> > What does `echo-*` test mean?
> 
> all the tests start with echo
> 
> ```
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_array_u8.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_big_int.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_big_uint.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_boxed_bytes.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_i32.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_i64.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_nothing.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_ser_ex_1.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_slice_u8.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_str_box.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_string.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_str.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_u64.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_usize.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_varags_tuples.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_varargs_u32.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_vec_u8.scen.json
> ```

Do you know what they do? They just output a value to the log?

File: elrond.md (hunk 2 above)
Comment (yzhang90) [reply elrond.md:980]: no. this error msg is hardcoded in the implementation of elrond. They will check this error msg in the output field. e.g., https://github.com/ElrondNetwork/elrond-wasm-rs/blob/bf438afe8216af459538565731ec8c76bf987d90/contracts/feature-tests/basic-features/mandos/echo_i32.scen.json#L111

File: elrond.md (hunk 3 above)
Comment (yzhang90) [reply elrond.md:986]: see above

**Comment by yzhang90** [comment]
> > What does `echo-*` test mean?
> 
> all the tests start with echo
> 
> ```
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_array_u8.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_big_int.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_big_uint.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_boxed_bytes.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_i32.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_i64.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_nothing.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_ser_ex_1.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_slice_u8.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_str_box.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_string.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_str.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_u64.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_usize.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_varags_tuples.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_varargs_u32.scen.json
> deps/elrond-wasm-rs/contracts/feature-tests/basic-features/mandos/echo_vec_u8.scen.json
> ```

read/write different types from/into memory and write into output cell.

File: Jenkinsfile (hunk 1 above)
Comment (yzhang90) [reply Jenkinsfile]: done

File: elrond.md (hunk 4 above)
Comment (yzhang90) [reply elrond.md:1021]: see above

File: elrond.md (hunk 2 above)
Comment (yzhang90) [reply elrond.md:980]: In the go implementation, the error msg is hard coded here https://github.com/ElrondNetwork/arwen-wasm-vm/blob/44d2b68c16a04266443f9dffc6ef21bda0b5d3cf/arwen/elrondapi/smallIntOps.go#L122

**Review by hjorthjort — APPROVED** [review]

**Pull request merged**